Extensions/Content/Dnn.PersonaBar.Extensions/WebApps/Pages.Web/src/components/dnn-persona-bar-page-treeview/src/PersonaBarPageTreeviewInteractor.jsx

```diff
@@ -664,6 +666,31 @@ class PersonaBarPageTreeviewInteractor extends Component {
                 });
             }
         });
+
+        const waitForChildList = (timeoutTime, callback) => new Promise((resolve, reject) => {
+            const check = () => {                
+                if(this.state.isChildLoaded) {
+                    callback();
+                    resolve();
+                } else if((timeoutTime -= 100) < 0) {
+                    reject();
+                } else {
+                    setTimeout(check, 100);
+                }                
+            }
+
+            setTimeout(check, 100)
+        });
+
+        if (this.props.multiplePagesAdded) {
+            this.props.addPagesComplete();
+            this.setState({ isChildLoaded: false });
+            this.loadAllChildList();
+
+            waitForChildList(2000, traverseCallback);
```

Comment:
Why not return the promise from loadAllChildList so you can use it like this: `this.loadAllChildList().then(traverseCallback);`

Reply:
There are many cascaded promises and asynchronous calls going down for `loadAllChildList`. In this case, we are waiting until the `isChildLoaded` state is set to `true`. Without this callback approach, there is a chance to miss this state transition. Hence, we are checking every 100 ms if this state is set to true or not up to to 2 seconds. Unfortunately the `promise.then` approach fails to deliver what is required in this specific case, otherwise you are right.